File: 0001-two-sum/0001-two-sum.java
import java.util.*;

class Solution {
    public int[] twoSum(int[] nums, int target) {
        //Array to store result
        int[] result = new int[2];
        //This map will store the difference and the corresponding index
        Map<Integer, Integer> map = new HashMap<>();
        //Traverse the entire array
        for(int i =0; i<nums.length; i++) {
            //if we seen the current element before then we have already encountered the other number of pair
            if(map.containsKey(nums[i])) {
                //index of the current element
                result[0] = i;
                //index of the other element of the pair
                result[1] = map.get(nums[i]);
            }
            //if we have not seen the current before, it means we have not yet encountered any number of the pair
             else {
                // Save the difference of the target and the current element
                // with the index of the current element
                map.put(target - nums[i], i);
            }
           
        }
         return result;
    }
}

File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(nums, target) {
    let n = nums.length;
    let eleIndex = [];

    // Step 1: store [value, index]
    for (let i = 0; i < n; i++) {
        eleIndex.push([nums[i], i]);
    }

    // Step 2: sort by value
    eleIndex.sort((a, b) => a[0] - b[0]);

    // Step 3: two-pointer approach
    let left = 0, right = n - 1;

    while (left < right) {
        let sum = eleIndex[left][0] + eleIndex[right][0];

        if (sum === target) {
            const i1 = eleIndex[left][1];
            const i2 = eleIndex[right][1];
            // Return in non-decreasing order
            return i1 < i2 ? [i1, i2] : [i2, i1];
        } else if (sum < target) {
            left++;
        } else {
            right--;
        }
    }

    return [-1, -1]; // fallback
};
